### PR TITLE
Add Weighted Sampler for highly imbalanced datasets

### PR DESCRIPTION
**Pull request by pourmand1376**
Related Issues:
- #492 
- #1115

This PR adds `weighted sampler` for datasets with highly imbalanced data. Idea is taken from [here](https://discuss.pytorch.org/t/how-to-handle-imbalanced-classes/11264). 

As you know in medical images, the data is mostly highly imbalanced and there is nothing we can do to increase data. If you train yolo using default sampler, you would get 0,0,0 for Precision, Recall and mAP. This is why this is a must for certain datasets. 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Added Weighted Sampler option for handling imbalanced datasets in YOLOv5 training.

### 📊 Key Changes
- Integrated Weighted Random Sampler within the training process.
- Added `weighted_sampler` as a boolean argument in the training script to enable the use of the sampler.
- Created a utility function to initialize the Weighted Random Sampler based on label distribution.
- Ensured the Weighted Sampler is not used during validation for accurate results.
  
### 🎯 Purpose & Impact
- 🎯 **Purpose:** To improve model performance on imbalanced datasets where some classes appear much more frequently than others.
- 💡 **Benefits:**
  - Helps to prevent model bias towards more frequent classes.
  - Aimed at boosting the accuracy for rare classes.
- ⚠️ **Potential Impact:**
  - Users training models on imbalanced datasets may see better performance.
  - The Weighted Sampler feature is not compatible with multi-GPU training setups at the moment.

### CODE REVIEW

**Comment by AyushExel** [comment]
@pourmand1376 hey this is a great idea. I've seen this being used for other tasks like segmentation. Do you have a before/after study of the effect of this change on any dataset?

**Comment by pourmand1376** [comment]
@AyushExel 
I am working on a paper which studies this effect on a custom dataset.  In the meantime, I will do an analysis on a public dataset and report here. Stay tuned.

**Comment by glenn-jocher** [comment]
@pourmand1376 very cool! Does this overlap or complement the train.py --image-weights argument which also introduces weighted sampling based on image contents and the previous epoch's per-class validation results?
https://github.com/ultralytics/yolov5/blob/e309a855860bc3f618c3541909c515a65ffc35b0/train.py#L466

**Comment by pourmand1376** [comment]
@glenn-jocher, I found no documentation for this option. I do not know what it does. Do we have any documentation about how to test it?

From the code, it seems that it does nothing serious. Am I right?

**Comment by triple-mu** [comment]
I test in coco128, there are something wrong in `unique_classes, counts = np.unique(labels_per_class, return_counts=True)`

**Comment by pourmand1376** [comment]
@triple-Mu  Thanks for testing! 

Can you explain more? How did you understand something is wrong?
Do you have any errors?

**Comment by pourmand1376** [comment]
I found [this](https://doc.bdd100k.com/download.html) public dataset to test my imbalanced sampling strategy. Although I have tested this on my custom dataset, testing should be done on public datasets to make it reproducible. 

I will do an analysis soon.

**Comment by triple-Mu** [comment]
> I found [this](https://doc.bdd100k.com/download.html) public dataset to test my imbalanced sampling strategy. Although I have tested this on my custom dataset, testing should be done on public datasets to make it reproducible.
> 
> I will do an analysis soon.

Coco128 is yolov5 default download dataset.
Command as such :
``` shell
python train.py --weighted_sampler
```
will reproduce my meeting error.

**Comment by pourmand1376** [comment]
@triple-Mu 
Thanks for reporting the bug. The bug that you reported is now solved. It can be easily tested via below codes in colab:

```bash
!git clone https://github.com/pourmand1376/yolov5
```
And the cell below:
```bash
cd yolov5
```

```bash
!git checkout add_sampler
!pip install -r requirements.txt
!python train.py --weighted_sampler
```

**Comment by triple-Mu** [comment]
> @triple-Mu 
> Thanks for reporting the bug. The bug that you reported is now solved. It can be easily tested via below codes in colab:
> 
> ```bash
> !git clone https://github.com/pourmand1376/yolov5
> ```
> ```bash
> cd yolov5
> ```
> ```bash
> !git checkout add_sampler
> !pip install -r requirements.txt
> !python train.py --weighted_sampler
> ```

Thanks a lot！

**Comment by glenn-jocher** [comment]
You're welcome, @triple-Mu! If you encounter any more issues or have further questions, feel free to reach out. Happy training! 😊🚀

File: utils/dataloaders.py [hunk 1]
```diff
@@ -100,6 +101,47 @@ def seed_worker(worker_id):
...
+    final_weights = []
+    for label in dataset.labels:
+        if label.shape[0] == 0:
+            final_weights.append(weight_background)
+        else:
+            # use weighted sum of labels for weight in case there are multiple labels for the same image
+            label_classes = np.unique(label[:, 0]).tolist()
+            values = []
+            for cls_ in label_classes:
+                values.append(weight_dict[_cls])
```
Comment (Ares-01) [inline utils/dataloaders.py:136]: Here shouldn't it be [cls_]? I assume this is a typo. Maybe worth noticing.

Comment (Ares-01) [inline utils/dataloaders.py:136]: Here shouldn't it be [cls_]? I assume this is a typo. Maybe worth noticing.

Comment (pourmand1376) [reply utils/dataloaders.py:136]: Yep. Well noticed. 

But this is not merged and it is not going to be since yolov5 is not maintained that much anymore.
